xarray/core/datatree.py

```diff
@@ -429,71 +527,87 @@ def parent(self: DataTree, new_parent: DataTree) -> None:
             raise ValueError("Cannot set an unnamed node as a child of another node")
         self._set_parent(new_parent, self.name)
 
+    def _to_dataset_view(self, rebuild_dims: bool) -> DatasetView:
+        variables = self._data_variables | self._coord_variables
+        if rebuild_dims:
+            dims = calculate_dimensions(variables)
+        else:
+            # Note: rebuild_dims=False can create technically invalid Dataset
+            # objects because it may not contain all dimensions on its direct
+            # member variables, e.g., consider:
+            #     tree = DataTree.from_dict(
+            #         {
+            #             "/": xr.Dataset({"a": (("x",), [1, 2])}),  # x has size 2
+            #             "/b/c": xr.Dataset({"d": (("x",), [3])}),  # x has size1
+            #         }
+            #     )
+            # However, they are fine for internal use cases, for align() or
+            # building a repr().
+            dims = dict(self._dims)
```

Comment:
@TomNicholas This is where we make the invalid Dataset objects, though they are also propagated by `_inherited_dataset()` (if one of the inputs is invalid).